### ISSUE TITLE
Storable: sv_upgrade from type 7 down to type 1

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#17037]
Migrated from [rt.perl.org#134179](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=134179) (status was 'pending release')

Searchable as RT134179$

**Comment by p5pRT** [comment]
### From alex.karelas@gmail.com

https://gist.github.com/akarelas-pt/f53e3f474e9cefbce8cf00af88f18c08

The above page demonstrates a Storable bug\.

I cannot find anywhere instructions on how to submit bugs for Storable\.

How do I submit bugs for Storable\, and\, more importantly\, how do I find instructions for how to submit bugs for Storable?

Thank you Sirs\,

\- Alexander

**Comment by p5pRT** [comment]
### From @ilmari

Hi Porters\,

daxim reported on IRC that the following program errors with $subject

&nbsp;    use Storable qw\(dclone\);
&nbsp;    use Scalar&#8203;::Util qw\(weaken\);

&nbsp;    my $foo = \[qr//\,\[\]\];
&nbsp;    weaken\($foo\->\[1\]\[0\]\[0\] = $foo\->\[1\]\);
&nbsp;    dclone $foo;

Both the preceding qr// and the two\-level weakened cycle are necessary
to trigger it\.

Backtrace&#8203;:

$ gdb \-q \-\-args perl \-MStorable=dclone \-MScalar&#8203;::Util=weaken \\
&nbsp;      \-e 'my $foo = \[qr//\,\[\]\]; weaken\($foo\->\[1\]\[0\]\[0\] = $foo\->\[1\]\); dclone $foo;'
Reading symbols from perl\.\.\.done\.
\(gdb\) b Perl\_croak
Function "Perl\_croak" not defined\.
Make breakpoint pending on future shared library load? \(y or \[n\]\) y
Breakpoint 1 \(Perl\_croak\) pending\.
\(gdb\) r
Starting program&#8203;: /home/ilmari/perl5/perlbrew/perls/30\.0td/bin/perl \-MStorable=dclone \-MScalar&#8203;::Util=weaken \-e my\\ \\$foo\\ =\\ \\\[qr//\,\\\[\\\]\\\]\\;\\ weaken\\\(\\$foo\-\\>\\\[1\\\]\\\[0\\\]\\\[0\\\]\\ =\\ \\$foo\-\\>\\\[1\\\]\\\)\\;\\ dclone\\ \\$foo\\;
\[Thread debugging using libthread\_db enabled\]
Using host libthread\_db library "/lib/x86\_64\-linux\-gnu/libthread\_db\.so\.1"\.

Breakpoint 1\, Perl\_croak \(my\_perl=my\_perl@&#8203;entry=0x555555757010\, pat=pat@&#8203;entry=0x7ffff7b49ee0 "sv\_upgrade from type %d down to type %d"\)
&nbsp;    at util\.c&#8203;:1753
1753    \{
\(gdb\) bt full
\#0  Perl\_croak \(my\_perl=my\_perl@&#8203;entry=0x555555757010\, pat=pat@&#8203;entry=0x7ffff7b49ee0 "sv\_upgrade from type %d down to type %d"\) at util\.c&#8203;:1753
&nbsp;        args = \{\{gp\_offset = 27\, fp\_offset = 0\, overflow\_arg\_area = 0x7ffff6206717 \<retrieve\+903>\, reg\_save\_area = 0x555555757010\}\}
\#1  0x00007ffff78d16e6 in Perl\_sv\_upgrade \(my\_perl=my\_perl@&#8203;entry=0x555555757010\, sv=sv@&#8203;entry=0x555555794858\, new\_type=new\_type@&#8203;entry=SVt\_IV\)
&nbsp;    at sv\.c&#8203;:1307
&nbsp;        old\_body = \<optimized out>
&nbsp;        new\_body = \<optimized out>
&nbsp;        old\_type = SVt\_PVMG
&nbsp;        new\_type\_details = \<optimized out>
&nbsp;        old\_type\_details = 0x7ffff7b56798 \<bodies\_by\_type\+56>
&nbsp;        referent = \<optimized out>
&nbsp;        \_\_PRETTY\_FUNCTION\_\_ = "Perl\_sv\_upgrade"
\#2  0x00007ffff620d0b5 in retrieve\_ref \(my\_perl=0x555555757010\, cxt=0x555555795378\, cname=0x0\) at Storable\.xs&#8203;:5435
&nbsp;        rv = 0x555555794858
&nbsp;        sv = 0x555555794450
&nbsp;        stash = 0x0
&nbsp;        \_\_PRETTY\_FUNCTION\_\_ = "retrieve\_ref"
\#3  0x00007ffff6206717 in retrieve \(my\_perl=my\_perl@&#8203;entry=0x555555757010\, cxt=cxt@&#8203;entry=0x555555795378\, cname=cname@&#8203;entry=0x0\)
&nbsp;    at Storable\.xs&#8203;:7350
&nbsp;        type = 4
&nbsp;        svh = \<optimized out>
&nbsp;        sv = \<optimized out>
&nbsp;        \_\_PRETTY\_FUNCTION\_\_ = "retrieve"
\#4  0x00007ffff6215735 in retrieve\_array \(my\_perl=0x555555757010\, cxt=0x555555795378\, cname=\<optimized out>\) at Storable\.xs&#8203;:6292
&nbsp;        len = 1
&nbsp;        i = 0
&nbsp;        av = 0x5555557887e8
&nbsp;        sv = \<optimized out>
&nbsp;        stash = \<optimized out>
&nbsp;        seen\_null = false
&nbsp;        \_\_PRETTY\_FUNCTION\_\_ = "retrieve\_array"
\#5  0x00007ffff6206717 in retrieve \(my\_perl=my\_perl@&#8203;entry=0x555555757010\, cxt=cxt@&#8203;entry=0x555555795378\, cname=cname@&#8203;entry=0x0\)
&nbsp;    at Storable\.xs&#8203;:7350
&nbsp;        type = 2
&nbsp;        svh = \<optimized out>
&nbsp;        sv = \<optimized out>
&nbsp;        \_\_PRETTY\_FUNCTION\_\_ = "retrieve"
\#6  0x00007ffff620ce34 in retrieve\_ref \(my\_perl=0x555555757010\, cxt=0x555555795378\, cname=0x0\) at Storable\.xs&#8203;:5410
&nbsp;        rv = 0x55555575a348
&nbsp;        sv = \<optimized out>
&nbsp;        stash = 0x0
&nbsp;        \_\_PRETTY\_FUNCTION\_\_ = "retrieve\_ref"
\#7  0x00007ffff6206717 in retrieve \(my\_perl=my\_perl@&#8203;entry=0x555555757010\, cxt=cxt@&#8203;entry=0x555555795378\, cname=cname@&#8203;entry=0x0\)
&nbsp;    at Storable\.xs&#8203;:7350
&nbsp;        type = 4
&nbsp;        svh = \<optimized out>
&nbsp;        sv = \<optimized out>
&nbsp;        \_\_PRETTY\_FUNCTION\_\_ = "retrieve"
\#8  0x00007ffff6215735 in retrieve\_array \(my\_perl=0x555555757010\, cxt=0x555555795378\, cname=\<optimized out>\) at Storable\.xs&#8203;:6292
&nbsp;        len = 2
&nbsp;        i = 1
&nbsp;        av = 0x55555577d520
&nbsp;        sv = \<optimized out>
&nbsp;        stash = \<optimized out>
&nbsp;        seen\_null = false
&nbsp;        \_\_PRETTY\_FUNCTION\_\_ = "retrieve\_array"
\#9  0x00007ffff6206717 in retrieve \(my\_perl=my\_perl@&#8203;entry=0x555555757010\, cxt=cxt@&#8203;entry=0x555555795378\, cname=cname@&#8203;entry=0x0\)
&nbsp;    at Storable\.xs&#8203;:7350
&nbsp;        type = 2
&nbsp;        svh = \<optimized out>
&nbsp;        sv = \<optimized out>
&nbsp;        \_\_PRETTY\_FUNCTION\_\_ = "retrieve"
\#10 0x00007ffff6207d1d in do\_retrieve \(my\_perl=my\_perl@&#8203;entry=0x555555757010\, f=f@&#8203;entry=0x0\, in=in@&#8203;entry=0x0\, optype=6\, optype@&#8203;entry=4\, 
&nbsp;    flags=flags@&#8203;entry=6\) at Storable\.xs&#8203;:7540
&nbsp;        perinterp\_sv = \<optimized out>
&nbsp;        cxt = 0x555555795378
&nbsp;        \_\_PRETTY\_FUNCTION\_\_ = "do\_retrieve"
&nbsp;        sv = \<optimized out>
&nbsp;        is\_tainted = \<optimized out>
&nbsp;        pre\_06\_fmt = 0
\#11 0x00007ffff620b60d in dclone \(sv=0x5555557888a8\, my\_perl=0x555555757010\) at Storable\.xs&#8203;:7741
&nbsp;        cxt = 0x555555795378
&nbsp;        real\_context = 0x555555795378
&nbsp;        perinterp\_sv = \<optimized out>
&nbsp;        size = 51
&nbsp;        out = \<optimized out>
\#12 XS\_Storable\_dclone \(my\_perl=0x555555757010\, cv=\<optimized out>\) at Storable\.xs&#8203;:7879
&nbsp;        sv = 0x5555557888a8
&nbsp;        RETVAL = \<optimized out>
&nbsp;        sp = \<optimized out>
&nbsp;        ax = 1
&nbsp;        mark = \<optimized out>
&nbsp;        items = \<optimized out>
\#13 0x00007ffff78b5236 in Perl\_pp\_entersub \(my\_perl=0x555555757010\) at pp\_hot\.c&#8203;:5237
&nbsp;        is\_scalar = \<optimized out>
&nbsp;        sp = \<optimized out>
&nbsp;        sv = 0x555555788d40
&nbsp;        gv = 0x7ffff78a3e87 \<Perl\_pp\_sassign\+119>
&nbsp;        cv = \<optimized out>
&nbsp;        old\_savestack\_ix = \<optimized out>
&nbsp;        \_\_PRETTY\_FUNCTION\_\_ = "Perl\_pp\_entersub"
\#14 0x00007ffff786d10a in Perl\_runops\_debug \(my\_perl=0x555555757010\) at dump\.c&#8203;:2537
&nbsp;        orig\_stack\_hwm = 0
\#15 0x00007ffff77ccb5b in S\_run\_body \(oldscope=1\, my\_perl=0x555555757010\) at perl\.c&#8203;:2716
No locals\.
\#16 perl\_run \(my\_perl=0x555555757010\) at perl\.c&#8203;:2639
&nbsp;        oldscope = 1
&nbsp;        ret = \<optimized out>
&nbsp;        cur\_env = \{je\_prev = 0x555555757320\, je\_buf = \{\{\_\_jmpbuf = \{1\, \-5698547751553798909\, 93824992235440\, 140737488346176\, 0\, 0\, 
&nbsp;                \-1891542260264669949\, \-1891523593628226301\}\, \_\_mask\_was\_saved = 0\, \_\_saved\_mask = \{\_\_val = \{93824994451896\, 93824994451944\, 
&nbsp;                  93824994451992\, 93824994452040\, 93824994452088\, 93824994452136\, 93824994452184\, 93824994340880\, 140737353934488\, 
&nbsp;                  993451154919501312\, 140737488346176\, 93824994340880\, 139637976727552\, 0\, 140737488346176\, 993451154919501312\}\}\}\}\, 
&nbsp;          je\_ret = 0\, je\_mustcatch = false\, je\_old\_delaymagic = 0\, je\_old\_stack\_hwm = 0\}
&nbsp;        \_\_PRETTY\_FUNCTION\_\_ = "perl\_run"
\#17 0x0000555555554f98 in main \(argc=\<optimized out>\, argv=\<optimized out>\, env=\<optimized out>\) at perlmain\.c&#8203;:127
&nbsp;        exitstatus = \<optimized out>
&nbsp;        i = \<optimized out>


Summary of my perl5 \(revision 5 version 30 subversion 0\) configuration&#8203;:
&nbsp;   
&nbsp;  Platform&#8203;:
&nbsp;    osname=linux
&nbsp;    osvers=4\.19\.0\-0\.bpo\.4\-amd64
&nbsp;    archname=x86\_64\-linux\-thread\-multi
&nbsp;    uname='linux ilmari\-thinkpad 4\.19\.0\-0\.bpo\.4\-amd64 \#1 smp debian 4\.19\.28\-2~bpo9\+1 \(2019\-03\-27\) x86\_64 gnulinux '
&nbsp;    config\_args='\-de \-Dprefix=/home/ilmari/perl5/perlbrew/perls/30\.0td \-Duseshrplib \-Dusethreads \-Dman1dir=none \-Dman3dir=none \-DDEBUGGING \-Aeval&#8203;:scriptdir=/home/ilmari/perl5/perlbrew/perls/30\.0td/bin'
&nbsp;    hint=recommended
&nbsp;    useposix=true
&nbsp;    d\_sigaction=define
&nbsp;    useithreads=define
&nbsp;    usemultiplicity=define
&nbsp;    use64bitint=define
&nbsp;    use64bitall=define
&nbsp;    uselongdouble=undef
&nbsp;    usemymalloc=n
&nbsp;    default\_inc\_excludes\_dot=define
&nbsp;    bincompat5005=undef
&nbsp;  Compiler&#8203;:
&nbsp;    cc='cc'
&nbsp;    ccflags ='\-D\_REENTRANT \-D\_GNU\_SOURCE \-fwrapv \-DDEBUGGING \-fno\-strict\-aliasing \-pipe \-fstack\-protector\-strong \-I/usr/local/include \-D\_LARGEFILE\_SOURCE \-D\_FILE\_OFFSET\_BITS=64 \-D\_FORTIFY\_SOURCE=2'
&nbsp;    optimize='\-O2 \-g'
&nbsp;    cppflags='\-D\_REENTRANT \-D\_GNU\_SOURCE \-fwrapv \-DDEBUGGING \-fno\-strict\-aliasing \-pipe \-fstack\-protector\-strong \-I/usr/local/include'
&nbsp;    ccversion=''
&nbsp;    gccversion='6\.3\.0 20170516'
&nbsp;    gccosandvers=''
&nbsp;    intsize=4
&nbsp;    longsize=8
&nbsp;    ptrsize=8
&nbsp;    doublesize=8
&nbsp;    byteorder=12345678
&nbsp;    doublekind=3
&nbsp;    d\_longlong=define
&nbsp;    longlongsize=8
&nbsp;    d\_longdbl=define
&nbsp;    longdblsize=16
&nbsp;    longdblkind=3
&nbsp;    ivtype='long'
&nbsp;    ivsize=8
&nbsp;    nvtype='double'
&nbsp;    nvsize=8
&nbsp;    Off\_t='off\_t'
&nbsp;    lseeksize=8
&nbsp;    alignbytes=8
&nbsp;    prototype=define
&nbsp;  Linker and Libraries&#8203;:
&nbsp;    ld='cc'
&nbsp;    ldflags =' \-fstack\-protector\-strong \-L/usr/local/lib'
&nbsp;    libpth=/usr/local/lib /usr/lib/gcc/x86\_64\-linux\-gnu/6/include\-fixed /usr/include/x86\_64\-linux\-gnu /usr/lib /lib/x86\_64\-linux\-gnu /lib/\.\./lib /usr/lib/x86\_64\-linux\-gnu /usr/lib/\.\./lib /lib
&nbsp;    libs=\-lpthread \-lnsl \-lgdbm \-ldb \-ldl \-lm \-lcrypt \-lutil \-lc \-lgdbm\_compat
&nbsp;    perllibs=\-lpthread \-lnsl \-ldl \-lm \-lcrypt \-lutil \-lc
&nbsp;    libc=libc\-2\.24\.so
&nbsp;    so=so
&nbsp;    useshrplib=true
&nbsp;    libperl=libperl\.so
&nbsp;    gnulibc\_version='2\.24'
&nbsp;  Dynamic Linking&#8203;:
&nbsp;    dlsrc=dl\_dlopen\.xs
&nbsp;    dlext=so
&nbsp;    d\_dlsymun=undef
&nbsp;    ccdlflags='\-Wl\,\-E \-Wl\,\-rpath\,/home/ilmari/perl5/perlbrew/perls/30\.0td/lib/5\.30\.0/x86\_64\-linux\-thread\-multi/CORE'
&nbsp;    cccdlflags='\-fPIC'
&nbsp;    lddlflags='\-shared \-O2 \-g \-L/usr/local/lib \-fstack\-protector\-strong'


Characteristics of this binary \(from libperl\)&#8203;: 
&nbsp;  Compile\-time options&#8203;:
&nbsp;    DEBUGGING
&nbsp;    HAS\_TIMES
&nbsp;    MULTIPLICITY
&nbsp;    PERLIO\_LAYERS
&nbsp;    PERL\_COPY\_ON\_WRITE
&nbsp;    PERL\_DONT\_CREATE\_GVSV
&nbsp;    PERL\_IMPLICIT\_CONTEXT
&nbsp;    PERL\_MALLOC\_WRAP
&nbsp;    PERL\_OP\_PARENT
&nbsp;    PERL\_PRESERVE\_IVUV
&nbsp;    PERL\_TRACK\_MEMPOOL
&nbsp;    USE\_64\_BIT\_ALL
&nbsp;    USE\_64\_BIT\_INT
&nbsp;    USE\_ITHREADS
&nbsp;    USE\_LARGE\_FILES
&nbsp;    USE\_LOCALE
&nbsp;    USE\_LOCALE\_COLLATE
&nbsp;    USE\_LOCALE\_CTYPE
&nbsp;    USE\_LOCALE\_NUMERIC
&nbsp;    USE\_LOCALE\_TIME
&nbsp;    USE\_PERLIO
&nbsp;    USE\_PERL\_ATOF
&nbsp;    USE\_REENTRANT\_API
&nbsp;    USE\_THREAD\_SAFE\_LOCALE
&nbsp;  Built under linux
&nbsp;  Compiled at May 24 2019 16&#8203;:22&#8203;:48
&nbsp;  %ENV&#8203;:
&nbsp;    PERL5LIB="/home/ilmari/\.perlbrew/libs/30\.0td@&#8203;std/lib/perl5"
&nbsp;    PERLBREW\_HOME="/home/ilmari/\.perlbrew"
&nbsp;    PERLBREW\_LIB="std"
&nbsp;    PERLBREW\_MANPATH="/home/ilmari/\.perlbrew/libs/30\.0td@&#8203;std/man&#8203;:/home/ilmari/perl5/perlbrew/perls/30\.0td/man"
&nbsp;    PERLBREW\_PATH="/home/ilmari/\.perlbrew/libs/30\.0td@&#8203;std/bin&#8203;:/home/ilmari/perl5/perlbrew/bin&#8203;:/home/ilmari/perl5/perlbrew/perls/30\.0td/bin"
&nbsp;    PERLBREW\_PERL="30\.0td"
&nbsp;    PERLBREW\_ROOT="/home/ilmari/perl5/perlbrew"
&nbsp;    PERLBREW\_SHELLRC\_VERSION="0\.82"
&nbsp;    PERLBREW\_VERSION="0\.84"
&nbsp;    PERL\_LOCAL\_LIB\_ROOT="/home/ilmari/\.perlbrew/libs/30\.0td@&#8203;std"
&nbsp;    PERL\_MB\_OPT="\-\-install\_base /home/ilmari/\.perlbrew/libs/30\.0td@&#8203;std"
&nbsp;    PERL\_MM\_OPT="INSTALL\_BASE=/home/ilmari/\.perlbrew/libs/30\.0td@&#8203;std"
&nbsp;  @&#8203;INC&#8203;:
&nbsp;    /home/ilmari/\.perlbrew/libs/30\.0td@&#8203;std/lib/perl5/x86\_64\-linux\-thread\-multi
&nbsp;    /home/ilmari/\.perlbrew/libs/30\.0td@&#8203;std/lib/perl5
&nbsp;    /home/ilmari/perl5/perlbrew/perls/30\.0td/lib/site\_perl/5\.30\.0/x86\_64\-linux\-thread\-multi
&nbsp;    /home/ilmari/perl5/perlbrew/perls/30\.0td/lib/site\_perl/5\.30\.0
&nbsp;    /home/ilmari/perl5/perlbrew/perls/30\.0td/lib/5\.30\.0/x86\_64\-linux\-thread\-multi
&nbsp;    /home/ilmari/perl5/perlbrew/perls/30\.0td/lib/5\.30\.0



\-\- 
"The surreality of the universe tends towards a maximum" \-\- Skud's Law
"Never formulate a law or axiom that you're not prepared to live with
 the consequences of\."                              \-\- Skud's Meta\-Law

**Comment by p5pRT** [comment]
### From @tonycoz

On Wed\, 05 Jun 2019 07&#8203;:41&#8203;:27 \-0700\, ilmari wrote&#8203;:
> Hi Porters\,
> 
> daxim reported on IRC that the following program errors with $subject
> 
> use Storable qw\(dclone\);
> use Scalar&#8203;::Util qw\(weaken\);
> 
> my $foo = \[qr//\,\[\]\];
> weaken\($foo\->\[1\]\[0\]\[0\] = $foo\->\[1\]\);
> dclone $foo;
> 

The attached fixes it for me\.

Tony

**Comment by p5pRT** [comment]
### From @tonycoz

<details><summary>134179-storable-regexp-self-ref.patch</summary>

```diff
From 911c22769d23520ac44e33511a52cd04341fba10 Mon Sep 17 00:00:00 2001
From: Tony Cook <tony@develop-help.com>
Date: Mon, 10 Jun 2019 10:17:20 +1000
Subject: (perl #134179) include regexps in the seen objects table on retrieve

Also, bless the regexp object, so freezing/thawing bless qr//, "Foo"
returns a "Foo" blesses regexp.
---
 dist/Storable/Storable.xs |  3 +++
 dist/Storable/t/regexp.t  |  4 +++-
 dist/Storable/t/weak.t    | 10 +++++++++-
 3 files changed, 15 insertions(+), 2 deletions(-)

diff --git a/dist/Storable/Storable.xs b/dist/Storable/Storable.xs
index ed729c94a6..ea794bbfe6 100644
--- a/dist/Storable/Storable.xs
+++ b/dist/Storable/Storable.xs
@@ -6808,6 +6808,7 @@ static SV *retrieve_regexp(pTHX_ stcxt_t *cxt, const char *cname) {
     SV *sv;
     dSP;
     I32 count;
+    HV *stash;
 
     PERL_UNUSED_ARG(cname);
 
@@ -6857,6 +6858,8 @@ static SV *retrieve_regexp(pTHX_ stcxt_t *cxt, const char *cname) {
 
     sv = SvRV(re_ref);
     SvREFCNT_inc(sv);
+    stash = cname ? gv_stashpv(cname, GV_ADD) : 0;
+    SEEN_NN(sv, stash, 0);
     
     FREETMPS;
     LEAVE;
diff --git a/dist/Storable/t/regexp.t b/dist/Storable/t/regexp.t
index acf28cfec6..e7c6c7e94a 100644
--- a/dist/Storable/t/regexp.t
+++ b/dist/Storable/t/regexp.t
@@ -37,7 +37,7 @@ while (<DATA>) {
     }
 }
 
-plan tests => 9 + 3*scalar(@tests);
+plan tests => 10 + 3*scalar(@tests);
 
 SKIP:
 {
@@ -75,6 +75,8 @@ SKIP:
     ok(!eval { dclone($re) }, "should fail to clone, even with use re 'eval'");
 }
 
+is(ref(dclone(bless qr//, "Foo")), "Foo", "check reblessed regexps");
+
 for my $test (@tests) {
     my ($code, $not, $match, $matchc, $name) = @$test;
     my $qr = eval $code;
diff --git a/dist/Storable/t/weak.t b/dist/Storable/t/weak.t
index 220c70160f..48752fbec4 100644
--- a/dist/Storable/t/weak.t
+++ b/dist/Storable/t/weak.t
@@ -29,7 +29,7 @@ sub BEGIN {
 }
 
 use Test::More 'no_plan';
-use Storable qw (store retrieve freeze thaw nstore nfreeze);
+use Storable qw (store retrieve freeze thaw nstore nfreeze dclone);
 require 'testlib.pl';
 our $file;
 use strict;
@@ -143,3 +143,11 @@ foreach (@tests) {
   $stored = nfreeze $input;
   tester($stored, \&freeze_and_thaw, $testsub, 'network string');
 }
+
+{
+    # [perl #134179] sv_upgrade from type 7 down to type 1
+    my $foo = [qr//,[]];
+    weaken($foo->[1][0][0] = $foo->[1]);
+    my $out = dclone($foo); # croaked here
+    is_deeply($out, $foo, "check they match");
+}
-- 
2.11.0


From 040de379ce9b7bec47a258d8b1d7449b21f70d2b Mon Sep 17 00:00:00 2001
From: Tony Cook <tony@develop-help.com>
Date: Mon, 10 Jun 2019 10:42:45 +1000
Subject: bump $Storable::VERSION

and update Changes
---
 dist/Storable/ChangeLog       | 5 +++++
 dist/Storable/__Storable__.pm | 2 +-
 2 files changed, 6 insertions(+), 1 deletion(-)

diff --git a/dist/Storable/ChangeLog b/dist/Storable/ChangeLog
index 04881992e2..4745c74b85 100644
--- a/dist/Storable/ChangeLog
+++ b/dist/Storable/ChangeLog
@@ -1,3 +1,8 @@
+2019-06-11 10:43:00 TonyC
+    version 3.16
+        * (perl #134179) fix self-referencing structures that include regexps
+        * bless regexps to preserve bless qr//, "Foo"
+
 2019-04-23 16:00:00 xsawyerx
     version 3.15
     * Fix leaking.
diff --git a/dist/Storable/__Storable__.pm b/dist/Storable/__Storable__.pm
index 9237371234..8ed247f96f 100644
--- a/dist/Storable/__Storable__.pm
+++ b/dist/Storable/__Storable__.pm
@@ -27,7 +27,7 @@ our @EXPORT_OK = qw(
 
 our ($canonical, $forgive_me);
 
-our $VERSION = '3.15';
+our $VERSION = '3.16';
 
 our $recursion_limit;
 our $recursion_limit_hash;
-- 
2.11.0


```
</details>

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @ilmari

"Tony Cook via RT" \<perlbug\-followup@&#8203;perl\.org> writes&#8203;:

> On Wed\, 05 Jun 2019 07&#8203;:41&#8203;:27 \-0700\, ilmari wrote&#8203;:
>> Hi Porters\,
>> 
>> daxim reported on IRC that the following program errors with $subject
>> 
>> use Storable qw\(dclone\);
>> use Scalar&#8203;::Util qw\(weaken\);
>> 
>> my $foo = \[qr//\,\[\]\];
>> weaken\($foo\->\[1\]\[0\]\[0\] = $foo\->\[1\]\);
>> dclone $foo;
>> 
>
> The attached fixes it for me\.

Looks good to me\, and matches how other potentially\-blessed refs are
handled\.

> diff \-\-git a/dist/Storable/Storable\.xs b/dist/Storable/Storable\.xs
> index ed729c94a6\.\.ea794bbfe6 100644
> \-\-\- a/dist/Storable/Storable\.xs
> \+\+\+ b/dist/Storable/Storable\.xs
> @&#8203;@&#8203; \-6808\,6 \+6808\,7 @&#8203;@&#8203; static SV \*retrieve\_regexp\(pTHX\_ stcxt\_t \*cxt\, const char \*cname\) \{
>      SV \*sv;
>      dSP;
>      I32 count;
> \+    HV \*stash;
>  
>      PERL\_UNUSED\_ARG\(cname\);

This is no longer unused\.

\- ilmari
\-\- 
"I use RMS as a guide in the same way that a boat captain would use
 a lighthouse\.  It's good to know where it is\, but you generally
 don't want to find yourself in the same spot\." \- Tollef Fog Heen

**Comment by p5pRT** [comment]
### From @jkeenan

On Tue\, 04 Jun 2019 20&#8203;:20&#8203;:41 GMT\, alex\.karelas@&#8203;gmail\.com wrote&#8203;:
> https://gist.github.com/akarelas-pt/f53e3f474e9cefbce8cf00af88f18c08
> 
> The above page demonstrates a Storable bug\.
> 
> I cannot find anywhere instructions on how to submit bugs for
> Storable\.
> 
> How do I submit bugs for Storable\, and\, more importantly\, how do I
> find instructions for how to submit bugs for Storable?
> 
> Thank you Sirs\,
> 
> \- Alexander

Your bug report was sent to 'perlbug\-admin@&#8203;perl\.org'\.  It should have been sent to 'perlbug@&#8203;perl\.org'\.

The sysadmins have moved it to the proper queue\.

Please include the string&#8203;:  \[perl \#134176\]
in the subject line of all future correspondence about this issue\.

Thank you very much\.

\-\- 
James E Keenan \(jkeenan@&#8203;cpan\.org\)

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @tonycoz

On Mon\, 10 Jun 2019 02&#8203;:53&#8203;:18 \-0700\, ilmari wrote&#8203;:
> "Tony Cook via RT" \<perlbug\-followup@&#8203;perl\.org> writes&#8203;:
> 
> > On Wed\, 05 Jun 2019 07&#8203;:41&#8203;:27 \-0700\, ilmari wrote&#8203;:
> >> Hi Porters\,
> >>
> >> daxim reported on IRC that the following program errors with
> >> $subject
> >>
> >> use Storable qw\(dclone\);
> >> use Scalar&#8203;::Util qw\(weaken\);
> >>
> >> my $foo = \[qr//\,\[\]\];
> >> weaken\($foo\->\[1\]\[0\]\[0\] = $foo\->\[1\]\);
> >> dclone $foo;
> >>
> >
> > The attached fixes it for me\.
> 
> Looks good to me\, and matches how other potentially\-blessed refs are
> handled\.
> 
> > diff \-\-git a/dist/Storable/Storable\.xs b/dist/Storable/Storable\.xs
> > index ed729c94a6\.\.ea794bbfe6 100644
> > \-\-\- a/dist/Storable/Storable\.xs
> > \+\+\+ b/dist/Storable/Storable\.xs
> > @&#8203;@&#8203; \-6808\,6 \+6808\,7 @&#8203;@&#8203; static SV \*retrieve\_regexp\(pTHX\_ stcxt\_t \*cxt\,
> > const char \*cname\) \{
> >      SV \*sv;
> >      dSP;
> >      I32 count;
> > \+    HV \*stash;
> >
> > PERL\_UNUSED\_ARG\(cname\);
> 
> This is no longer unused\.

Thanks\, fixed in the version committed\.

The fix itself is 16f2ddb794883529d5a3ad8326974a07aae7e567 and the version bump is 5afa8ffbdf01228399875ca4111e261e206f40cd\.

This fixes both this ticket and the DateTime&#8203;::Format&#8203;::Strptime issue now merged into 134179 \(since the DateTime&#8203;::Format&#8203;::Strptime object has a regular expression embedded in it\.\)

Tony

**Comment by p5pRT** [comment]
@tonycoz - Status changed from 'open' to 'pending release'

**Comment by p5pRT** [comment]
### From alex.karelas@gmail.com

Is it my fault I sent it there? That's the instructions on the screen 
said\. \(I reached them by clicking on 'Issues' on the left\-hand side on 
metacpan's page for Storable\)

On 10/6/19 7&#8203;:33 μ\.μ\.\, James E Keenan via RT wrote&#8203;:
> On Tue\, 04 Jun 2019 20&#8203;:20&#8203;:41 GMT\, alex\.karelas@&#8203;gmail\.com wrote&#8203;:
>> https://gist.github.com/akarelas-pt/f53e3f474e9cefbce8cf00af88f18c08
>>
>> The above page demonstrates a Storable bug\.
>>
>> I cannot find anywhere instructions on how to submit bugs for
>> Storable\.
>>
>> How do I submit bugs for Storable\, and\, more importantly\, how do I
>> find instructions for how to submit bugs for Storable?
>>
>> Thank you Sirs\,
>>
>> \- Alexander
> Your bug report was sent to 'perlbug\-admin@&#8203;perl\.org'\.  It should have been sent to 'perlbug@&#8203;perl\.org'\.
>
> The sysadmins have moved it to the proper queue\.
>
> Please include the string&#8203;:  \[perl \#134176\]
> in the subject line of all future correspondence about this issue\.
>
> Thank you very much\.
>
